File: Astro-TFG/public/scripts/4.1-calculos-matematicos/prueba-3/version-2/javascript/main.js
let highPrecisionMath;

function initHighPrecisionMath() {
    const mathLib = typeof math !== 'undefined' ? math : mathjs;
    const { create, all } = mathLib;

    highPrecisionMath = create(all, {
        number: 'BigNumber',
        precision: 10000
    });
}

function calculatePiGaussLegendre(digits) {
    initHighPrecisionMath()
    const math = highPrecisionMath;

    math.config({ precision: digits });

    const one = math.bignumber(1);
    const half = math.bignumber(0.5);
    const quarter = math.bignumber(0.25);
    const two = math.bignumber(2);

    let a = one;
    let b = math.sqrt(half);
    let t = quarter;
    let p = one;

    const numIter = Math.max(5, Math.floor(Math.log10(digits) * 1.4));

    for (let i = 0; i < numIter; i++) {
        const aNext = math.divide(math.add(a, b), two);
        b = math.sqrt(math.multiply(a, b));
        const diff = math.subtract(a, aNext);
        t = math.subtract(t, math.multiply(p, math.pow(diff, two)));
        a = aNext;
        p = math.multiply(p, two);
    }

    const pi = math.divide(math.pow(math.add(a, b), 2), math.multiply(4, t));
    return pi.toString();
}

async function n_digits_pi(repetitions, digits) {
    const outputDiv = document.getElementById("javascript-output");

    let totalTime = 0;
    let totalMemory = 0;
    const startTotal = performance.now();

    for (let i = 0; i < repetitions; i++) {
        const start = performance.now();

        const piValue = calculatePiGaussLegendre(digits);

        const end = performance.now();
        const memoryUsage = performance.memory ?
            (performance.memory.usedJSHeapSize / (1024 * 1024)) : 0;

        totalTime += (end - start);
        totalMemory += memoryUsage;
    }

    // ET (Execution Time)

    const endTotal = performance.now();
    const totalExecTime = (endTotal - startTotal).toFixed(2);
    const avgTime = (totalTime / repetitions).toFixed(2);

    // RAM

    const avgMemory = (totalMemory / repetitions).toFixed(2);

    outputDiv.innerHTML = `
        <div>Total ET (10x): ${totalExecTime} ms</div>
        <div>ET (avg, 10x): ${avgTime} ms</div>
        <div>RAM (avg, 10x): ${avgMemory} MB</div>
    `;
}

window.runJSBenchmark = function () {
    clearCell("javascript-output");
    window.showExecutionLoader();

    requestAnimationFrame(() => {
        setTimeout(() => {
            n_digits_pi(10, 10_000);
            window.hideExecutionLoader();
        }, 0);
    });

}

function clearCell(elementId) {
    document.getElementById(elementId).innerHTML = "";
}

File: Astro-TFG/public/scripts/4.1-calculos-matematicos/prueba-3/version-2/python/main.py
import gc
import time
import tracemalloc
import js  # type: ignore
from pyscript import display  # type: ignore
from mpmath import mp


def calculate_pi_gauss_legendre(digits):
    mp.dps = digits

    a = mp.mpf(1)
    b = 1 / mp.sqrt(2)
    t = mp.mpf(1) / 4
    p = mp.mpf(1)

    num_iter = int(mp.log10(digits) * 1.4)

    for _ in range(num_iter):
        a_next = (a + b) / 2
        b = mp.sqrt(a * b)
        t -= p * (a - a_next) ** 2
        a = a_next
        p *= 2

    pi = ((a + b) ** 2) / (4 * t)
    return pi


def n_digits_pi(repetitions, digits):
    total_time = 0
    total_memory = 0

    start_total = time.time()

    for _ in range(repetitions):
        tracemalloc.start()

        start = time.time()
        gc.collect()
        pi_value = calculate_pi_gauss_legendre(digits)
        end = time.time()

        memory_usage = tracemalloc.get_traced_memory()[1] / (1024 * 1024)
        tracemalloc.stop()

        total_time += (end - start) * 1000
        total_memory += memory_usage

    # ET (Execution Time)

    end_total = time.time()
    total_exec_time = round((end_total - start_total) * 1000, 2)
    avg_time = round(total_time / repetitions, 2)

    # RAM

    avg_memory = round(total_memory / repetitions, 2)

    display(f"Total ET (10x): {total_exec_time} ms",
            target="pyscript-output")
    display(f"ET (avg, 10x): {avg_time} ms", target="pyscript-output")
    display(f"RAM (avg, 10x): {avg_memory} MB", target="pyscript-output")
    js.endTimerWebAssembly()


def run_py_benchmark(event):
    js.clearCell('pyscript-output')
    n_digits_pi(10, 10_000)


js.window.run_py_benchmark = run_py_benchmark
